cpr_reputation/utils.py

```diff
@@ -143,11 +193,118 @@ def get_config(
         ],
     }
 
-    return env_config, ray_config, heterogenous
+    return env_config, ray_config, run_config
+
+
+# ef gini(rewards: Sequence[float]) -> float:
+#    """https://en.wikipedia.org/wiki/Gini_coefficient#Calculation"""
+#    return 1 - sum(sum(abs(ri - rk) for ri, rk in product(rewards, rewards))) / 2 / len(
+#        rewards
+#    ) / sum(rewards)
+
+
+class CPRCallbacks(DefaultCallbacks):
+    def __init__(self):
+        super().__init__()
+
+    def on_episode_start(
+        self,
+        *,
+        worker: "RolloutWorker",
+        base_env: BaseEnv,
+        policies: Dict[PolicyID, Policy],
+        episode: MultiAgentEpisode,
+        env_index: Optional[int] = None,
+        **kwargs,
+    ) -> None:
+        episode.user_data["rewards"] = []
+        episode.user_data["rewards_gini"] = []
+        episode.user_data["reputations"] = []
+        episode.user_data["reputations_gini"] = []
+        episode.user_data["num_shots"] = []
+        episode.user_data["sustainability"] = []
+
+    def on_episode_step(
+        self,
+        *,
+        worker: "RolloutWorker",
+        base_env: BaseEnv,
+        episode: MultiAgentEpisode,
+        env_index: Optional[int] = None,
+        **kwargs,
+    ) -> None:
+        num_shots = {}
+        rewards = {}
+        for agent_id, _ in base_env.get_unwrapped()[0].game.agents.items():
+            num_shots[agent_id] = episode.last_action_for(agent_id) == SHOOT
+            rewards[agent_id] = episode.prev_reward_for(agent_id)
+        reputations = base_env.get_unwrapped()[0].game.reputation
+        episode.user_data["rewards"].append(rewards)
+        episode.user_data["rewards_gini"].append(gini_coef(rewards))
+        episode.user_data["reputations"].append(reputations)
+        episode.user_data["reputations_gini"].append(gini_coef(reputations))
+        episode.user_data["num_shots"].append(num_shots)
+        sus_metric = sustainability_metric(
+            base_env.get_unwrapped()[0].game.sustainability_metric
+        )
+        episode.user_data["sustainability"].append(
+            sus_metric(episode.user_data["rewards"])
+        )
+
+    def on_episode_end(
+        self,
+        *,
+        worker: "RolloutWorker",
+        base_env: BaseEnv,
+        policies: Dict[PolicyID, Policy],
+        episode: MultiAgentEpisode,
+        env_index: Optional[int] = None,
+        **kwargs,
+    ) -> None:
+        # custom metrics get saved to the logfile
+        episode.custom_metrics["rewards"] = (
+            sum(
```

Comment:
@bengreenberg5 can you explain to me why "rewards" is the sum over the sum over all the intermediate rewards while rewards_gini is just from the last episode?

Reply:
Yeah you're right that it's convoluted. `rewards` keeps track of the reward scalar for all agents; `rewards_gini` is a simple scalar metric. `rewards` is a `List[Dict[str, float]]` (rewards per agent, per episode) and `rewards_gini` is a `List[float]` (inequality of agents' rewards, per episode).

Reply:
> inequality of agents' rewards, per episode

I think you mean "per step"? anyway, same difference. I made sure it's inequality of _accumulated_ rewards instead of one step's rewards.